### ISSUE TITLE
Support for metrics for a service deployed with multiple JVMs

### ISSUE THREAD

**Issue by michaelhyatt** [elastic/apm-agent-java#826]
Applications can be deployed across multiple hosts/docker containers, but have the same service name. Currently, all the metrics are reported for the same service name without the ability to see how individual instances of the deployed service perform. I'd like to see a service instance identifier field (similar to hostname) to be able to distinguish between the metrics coming from different instances. Also, UI support in Kibana is needed.

**Comment by axw** [comment]
This seems relevant: https://github.com/elastic/kibana/issues/43765

@felixbarny @eyalkoren is there anything to do in the agent for this?

**Comment by michaelhyatt** [comment]
adding @bansriyar

I don't see any fields here capturing something like hostname/instance name
https://www.elastic.co/guide/en/apm/agent/java/current/metrics.html

**Comment by axw** [comment]
@michaelhyatt there's a standard set of metadata that gets added to all of the documents, including metrics: https://www.elastic.co/guide/en/apm/server/current/metadata-api.html. Take a look at the metrics docs in [demo.elastic.co](https://demo.elastic.co/app/kibana#/discover?_g=()&_a=(columns:!(host.hostname),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:'apm-*',key:processor.event,negate:!f,params:(query:metric),type:phrase,value:metric),query:(match:(processor.event:(query:metric,type:phrase))))),index:'apm-*',interval:auto,query:(language:kuery,query:''),sort:!('@timestamp',desc))).

**Comment by eyalkoren** [comment]
This is related to an earlier discussion leading to the Kibana issue linked above by @axw - if you look at the referred prototype in that same issue you'd see that service instances are already top-level citizens in the metrics landing page. The question is- where these JVM identifiers come from?

There was a discussion preceding this issue (see part of it starting in https://github.com/elastic/kibana/issues/43009#issuecomment-520212246). Essentially, it would be a combination of host-environment and/or container-id, and I think there was an agreement to add a configuration option for setting an instance ID manually wherever those are insufficient.
Note that this is also relevant for other data you may want to filter by JVM, so this also applies to the drop-down filters we want to have.

> is there anything to do in the agent for this?

A possible future enhancement would be to "discover" a meaningful instance unique ID, like host:port, but I don't think this is something we are going to have an open issue for at the moment.

So, basically, all we need to do in this regard is add this manually configurable instance ID, if it is decided to go for it.

Therefore, I believe we can close this issue, unless @michaelhyatt feels it is not addressed?

**Comment by michaelhyatt** [comment]
I can see there is `host.hostname` and `container.id` at the moment, so it needs to be supported in Kibana to have an end to end working solution. Feel free to close it, as long as the instance.id enhancement is tracked elsewhere as a low priority ticket.

**Comment by eyalkoren** [comment]
I actually think the `instance.id` should be high priority 🙂 
Anyway, it first needs to be decided and then opened as an issue in the APM repo to be implemented by all agents and server.
So I am closing this one.
